### PR TITLE
feat(byods-sdk):added logger package to base-client,token-manager,byd…

### PR DESCRIPTION
**Pull request by RAM1232**
…os and data-source-client

# COMPLETES #<[SPARK-561702](https://jira-eng-gpk2.cisco.com/jira/browse/SPARK-561702)>

## This pull request addresses

This pull request addresses the addition of logger and error packages to the BYODOS package, replacing the console messages with log messages.

## by making the following changes
By adding the _BYODOSDeviceError.ts_ package and the _ExtendError_ package in the BYDOS/Error folder, along with _index.ts_ and _type.ts_ in the _BYODOS/LOGGER_ folder, references to these were included in _BYODOS/base-client_, _BYODOS/datasource-client_, _BYODOS/byodos_, and _BYODOS/token-manager_. This implementation adds logger packages to BYODOS and replaces the console messages with logger messages.

### Change Type

- [ ] Bug fix (non-breaking change which fixes an issue)
- [ ] New feature (non-breaking change which adds functionality)
- [ ] Breaking change (fix or feature that would cause existing functionality to change)
- [ ] Documentation update
- [ ] Tooling change
- [ ] Internal code refactor

## The following scenarios where tested
The Packages are builded without any errors with all test cases were passed.

### I certified that

- [ ] I have read and followed [contributing guidelines](https://github.com/webex/webex-js-sdk/blob/master/CONTRIBUTING.md#submitting-a-pull-request)
- [ ] I discussed changes with code owners prior to submitting this pull request

- [ ] I have not skipped any automated checks
- [ ] All existing and new tests passed
- [ ] I have updated the documentation accordingly

---

Make sure to have followed the [contributing guidelines](https://github.com/webex/webex-js-sdk/blob/master/CONTRIBUTING.md#submitting-a-pull-request) before submitting.

### CODE REVIEW

File: packages/byods/src/Errors/catalog/BYODSDeviceError.ts [hunk 1]
```diff
@@ -0,0 +1,70 @@
...
+//import {RegistrationStatus} from '../../common/types';
+
+import {ErrorContext, ErrorMessage, ErrorObject, ERROR_TYPE} from '../types';
+import ExtendedError from './ExtendedError';
+
+/**
+ * Any error reported from Calling client should be stored here.
+ */
+export class BYODSError extends ExtendedError {
+  // public status: RegistrationStatus = RegistrationStatus.INACTIVE;
```
Comment (adhmenon) [inline packages/byods/src/Errors/catalog/BYODSDeviceError.ts:11]: Do we need this comment? I notice a lot of comments in the file. Can we remove the ones which are not required?

**Comment by adhmenon** [comment]
@RAM1232 , please do add the details like JIRA ticket and explanation.

File: packages/byods/src/Logger/types.ts [hunk 2]
```diff
@@ -0,0 +1,33 @@
...
+// import {IMetaContext} from '../common/types';
+export interface IMetaContext {
+  file?: string;
+  method?: string;
+}
+
+export interface LogContext extends IMetaContext {}
+
+export enum LOG_PREFIX {
+  MAIN = 'CALLING_SDK',
```
Comment (adhmenon) [inline packages/byods/src/Logger/types.ts:11]: Can we rename this to BYODS?

File: packages/byods/src/base-client/type.ts [hunk 3]
```diff
@@ -0,0 +1,107 @@
...
+  logger?: LoggerConfig;
+}
+
+export type BYODSErrorEmitterCallback = (
+  err: BYODSError,
+  finalError?: boolean
+) => void;
+
+/**
+ * An interface for the `CallingClient` module.
```
Comment (adhmenon) [inline packages/byods/src/base-client/type.ts:18]: Nit: I think all this documentation is for the calling sdk, can we remove it and tailor it for the current byods sdk.

File: packages/byods/src/Errors/types.ts [hunk 4]
```diff
@@ -0,0 +1,60 @@
...
+
+export type ErrorMessage = string;
+
+export enum ERROR_LAYER {
+  CALL_CONTROL = 'call_control',
+  MEDIA = 'media',
+}
+
+export enum ERROR_TYPE {
+  CALL_ERROR = 'call_error',
```
Comment (adhmenon) [inline packages/byods/src/Errors/types.ts:11]: Don't think we need `CALL_ERROR` here - correct me if I'm wrong.

File: packages/byods/src/Errors/types.ts [hunk 5]
```diff
@@ -0,0 +1,60 @@
...
+  ERROR_PROCESSING = 114,
+  USER_BUSY = 115,
+  PARSING_ERROR = 116,
+  TIMEOUT_ERROR = 117,
+  NOT_ACCEPTABLE = 118,
+  CALL_REJECTED = 119,
+  NOT_AVAILABLE = 120,
+}
+
+export enum DEVICE_ERROR_CODE {
```
Comment (adhmenon) [inline packages/byods/src/Errors/types.ts:47]: I don't think we need any of these - devices are seen in calling and meeting sdk - byods will be a backend sdk so we would not require it here.

File: packages/byods/src/Errors/types.ts [hunk 6]
```diff
@@ -0,0 +1,60 @@
...
+  DEVICE_NOT_FOUND = 404,
+  INTERNAL_SERVER_ERROR = 500,
+  NOT_IMPLEMENTED = 501,
+  SERVICE_UNAVAILABLE = 503,
+  BAD_REQUEST = 400,
+  REQUEST_TIMEOUT = 408,
+  TOO_MANY_REQUESTS = 429,
+}
+
+export enum CALL_ERROR_CODE {
```
Comment (adhmenon) [inline packages/byods/src/Errors/types.ts:34]: Same over here... we can remove this and only keep errors pertaining to the current SDK in the code.

**Review by adhmenon — CHANGES_REQUESTED** [review]
Please do make the required changes and remove the code which is related to the calling sdk errors. Most of them should not be required.

File: packages/byods/src/Errors/catalog/BYODSDeviceError.ts (hunk 1 above)
Comment (RAM1232) [reply packages/byods/src/Errors/catalog/BYODSDeviceError.ts]: @adhmenon Updated...!

File: packages/byods/src/Logger/types.ts (hunk 2 above)
Comment (RAM1232) [reply packages/byods/src/Logger/types.ts]: @adhmenon Updated...!

File: packages/byods/src/base-client/type.ts (hunk 3 above)
Comment (RAM1232) [reply packages/byods/src/base-client/type.ts]: @adhmenon Updated...!

File: packages/byods/src/Errors/types.ts (hunk 4 above)
Comment (RAM1232) [reply packages/byods/src/Errors/types.ts]: @adhmenon Updated...!

File: packages/byods/src/Errors/types.ts (hunk 5 above)
Comment (RAM1232) [reply packages/byods/src/Errors/types.ts]: @adhmenon Updated...!

File: packages/byods/src/Errors/types.ts (hunk 6 above)
Comment (RAM1232) [reply packages/byods/src/Errors/types.ts]: @adhmenon Updated...!

File: packages/byods/src/Errors/catalog/BYODSDeviceError.ts [hunk 7]
```diff
@@ -0,0 +1,36 @@
+/* eslint-disable valid-jsdoc */
```
Comment (Shreyas281299) [inline packages/byods/src/Errors/catalog/BYODSDeviceError.ts:1]: Do we need this? What error are we getting if we don't have this line?

File: packages/byods/src/byods/constant.ts [hunk 8]
```diff
@@ -0,0 +1 @@
+export const BYODS_FILE = 'byods';
```
Comment (Shreyas281299) [inline packages/byods/src/byods/constant.ts:1]:
```suggestion
export const BYODS = 'byods';
```

File: packages/byods/src/byods/index.ts [hunk 9]
```diff
@@ -10,6 +10,10 @@ import {
 } from '../constants';
 import {SDKConfig, JWSTokenVerificationResult} from '../types';
 import TokenManager from '../token-manager';
+import {BYODSConfig} from '../token-manager/type';
```
Comment (Shreyas281299) [inline packages/byods/src/byods/index.ts:13]: In BYODS class we are using types from token-mnager, We can either use the types that is created in byods class or we can have a common types file.

File: packages/byods/src/byods/index.ts [hunk 10]
```diff
@@ -37,9 +42,12 @@ export default class BYODS {
    * @example
    * const sdk = new BYODS({ clientId: 'your-client-id', clientSecret: 'your-client-secret' });
    */
-  constructor({clientId, clientSecret}: SDKConfig) {
+  constructor({clientId, clientSecret}: SDKConfig, config?: BYODSConfig) {
```
Comment (Shreyas281299) [inline packages/byods/src/byods/index.ts:45]: Do you think it would be better to add logger config in SDKConfig rather than creating a new config for it,
```suggestion
  constructor({clientId, clientSecret, logger}: SDKConfig) {
```

File: packages/byods/src/byods/type.ts
Comment (Shreyas281299) [inline packages/byods/src/byods/type.ts:1]: We have the same interfaces defined in 4 places. I think we can have a types file at the root and then use them.

File: packages/byods/src/data-source-client/constants.ts [hunk 11]
```diff
@@ -1 +1,2 @@
 export const DATASOURCE_ENDPOINT = '/dataSources';
+export const BYODS_DATA_SOURCE_CLIENT_FILE = 'data-source-client';
```
Comment (Shreyas281299) [inline packages/byods/src/data-source-client/constants.ts:2]:
```suggestion
export const BYODS_DATA_SOURCE_CLIENT = 'data-source-client';
```

File: packages/byods/src/data-source-client/types.ts [hunk 12]
```diff
@@ -1,63 +1,40 @@
...
+  schemaId: string; // The identifier for the schema associated with the data source.
+  orgId: string; // The identifier for the organization associated with the data source.
+  applicationId: string; // The plain client identifier (not a Hydra base64 id string).
+  status: string; // The status of the data source response. Either "active" or "disabled".
+  jwsToken: string; // The JSON Web Signature token associated with the data source response.
+  createdBy: string; // The identifier of the user who created the data source response.
+  createdAt: string; // The timestamp when the data source response was created.
+  updatedBy?: string; // The identifier of the user who last updated the data source response.
+  updatedAt?: string; // The timestamp when the data source response was last updated.
+  errorMessage?: string; // The error message associated with the data source response, if any.
```
Comment (Shreyas281299) [inline packages/byods/src/data-source-client/types.ts:37]: Please revert these changes as they are necessary to create the type doc

File: packages/byods/src/data-source-client/types.ts [hunk 13]
```diff
@@ -66,33 +43,10 @@ export interface DataSourceResponse {
...
-  /**
-   * The lifetime of the token in minutes.
-   */
-  tokenLifetimeMinutes: number;
+  schemaId: string; // The identifier for the schema associated with the data source.
+  url: string; // The URL of the data source.
+  audience: string; // The audience for the data source request.
+  subject: string; // The subject of the data source request.
+  nonce: string; // A unique nonce for the data source request.
+  tokenLifetimeMinutes: number; // The lifetime of the token in minutes.
```
Comment (Shreyas281299) [inline packages/byods/src/data-source-client/types.ts:51]: Same as above

File: packages/byods/src/token-manager/constant.ts [hunk 14]
```diff
@@ -0,0 +1,2 @@
+const BYODS_TOKEN_MANAGER_FILE = 'token-manager';
```
Comment (Shreyas281299) [inline packages/byods/src/token-manager/constant.ts:1]:
```suggestion
const BYODS_TOKEN_MANAGER = 'token-manager';
```

File: packages/byods/src/token-manager/index.ts [hunk 15]
```diff
@@ -21,14 +28,22 @@ export default class TokenManager {
    * @example
    * const tokenManager = new TokenManager('your-client-id', 'your-client-secret');
    */
-  constructor(clientId: string, clientSecret: string, baseUrl: string = PRODUCTION_BASE_URL) {
+  constructor(
+    clientId: string,
+    clientSecret: string,
+    config?: BYODSConfig,
```
Comment (Shreyas281299) [inline packages/byods/src/token-manager/index.ts:34]: Same here, would it be better to pass the level directly rather than the whole config object

File: packages/byods/src/token-manager/index.ts [hunk 16]
```diff
@@ -161,12 +190,21 @@ export default class TokenManager {
...
+        }),
       });
 
-      this.updateServiceAppToken(response.data, orgId);
+      if (!response.ok) {
+        throw new Error(`Failed to save service app registration: ${response.statusText}`);
+      }
+
+      const data: TokenResponse = (await response.json()) as TokenResponse;
+      this.updateServiceAppToken(data, orgId);
```
Comment (Shreyas281299) [inline packages/byods/src/token-manager/index.ts:201]: Please correct me if Im wrong but I think these changes are from a different PR.

File: packages/byods/test/unit/spec/Logger/index.text.ts
Comment (Shreyas281299) [inline packages/byods/test/unit/spec/Logger/index.text.ts:1]: There is a spelling mistake in the file name.

File: packages/byods/test/unit/spec/Logger/index.text.ts [hunk 17]
```diff
@@ -0,0 +1,77 @@
...
+
+    log.trace(fakePrint, dummyContext);
+    expect(traceSpy).not.toHaveBeenCalledTimes(1);
+  });
+
+  it('Set the log level to Trace  and verify that all levels are executed', () => {
+    log.setLogger(LOGGER.TRACE);
+
+    log.info(fakePrint, dummyContext);
+    expect(logSpy).toHaveBeenCalledTimes(2); // one during initialization and one with the statement
```
Comment (Shreyas281299) [inline packages/byods/test/unit/spec/Logger/index.text.ts:66]: Can we also check if the correct set of strings are printed.

**Review by Shreyas281299 — CHANGES_REQUESTED** [review]
Please see the comments
